File: 프로그래머스/l1_체육복.py
def solution(n, lost, reserve):
    real_lost = set(lost) - set(reserve)
    real_reserve = set(reserve) - set(lost)
    cnt = 0

    for l in real_lost:
        if l-1 in real_reserve:
            cnt +=1

            real_reserve.remove(l-1)
        elif l+1 in real_reserve:
            cnt+=1
            real_reserve.remove(l+1)

    return n - len(real_lost) + cnt


solution(5, [2,4], [1,3,5])
solution(5, [2,4], [3])
solution(3, [3], [1])


# def solution(n, lost, reserve):
#     r_lost = set(lost) - set(reserve)
#     r_reserve = set(reserve) - set(lost)
    
#     cnt = 0
#     for l in r_lost:
#         if l-1 in r_reserve:
#             r_reserve.remove(l-1)
#             cnt+=1
#         elif l+1 in r_reserve:
#             r_reserve.remove(l+1)
#             cnt+=1
#     # print(n, r_lost, r_reserve, cnt)
#     # print(len(r_lost))
#     return n - len(r_lost) + cnt
#     # for r in r_reserve:
#     #     if r-1 in r_lost:
#     #         r_lost.remove(r-1)
#     #     elif r+1 in r_lost:
#     #         r_lost.remove(r+1)

File: 프로그래머스_JAVA/l1_체육복.java
import java.util.HashSet;

public class l1_체육복 {

    public static void main(String[] args){
        int a = 5;
        int[] b = {2,4};
        int[] c = {1,3,5};
        int d = solution(a,b,c);
    }
    public static int solution(int n, int[] lost, int[] reserve) {
        int answer = n -lost.length;
        HashSet<Integer> realReserve = new HashSet<Integer>();
        for (int r:reserve){
            realReserve.add(r);
        }
        for(int i=0; i<lost.length; i++){
            if(realReserve.contains(lost[i])) {
                answer++;
                realReserve.remove(lost[i]);
                lost[i] = -1;
            }
        }
        for (int i=0; i<lost.length;i++){
            if(realReserve.contains(lost[i]-1)){
                answer++;
                realReserve.remove(lost[i]-1);
            }else if(realReserve.contains(lost[i]+1)){
                answer++;
                realReserve.remove(lost[i]+1);
            }
        }

        return answer;
    }
}
